scripts/transformer_qa_eval.py

```diff
@@ -0,0 +1,85 @@
+import json
+import logging
+import time
+from typing import Iterable, List
+
+from allennlp.common.checks import check_for_gpu
+from allennlp.data import Instance
+from allennlp.predictors import Predictor
+
+from allennlp_rc.eval import SquadEmAndF1
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+if __name__ == "__main__":
+    import allennlp_rc  # noqa F401: Needed to register the registrables.
+    import argparse
+
+    logging.basicConfig(level=logging.INFO)
+
+    parser = argparse.ArgumentParser(description="Evaluation for SQuAD 1.1")
+    parser.add_argument("--cuda-device", type=int, default=-1)
+    parser.add_argument("--qa-model", type=str)
+    parser.add_argument("--input-file", type=str)
+    args = parser.parse_args()
+
+    # Read inputs
+    check_for_gpu(args.cuda_device)
+    predictor = Predictor.from_path(
+        args.qa_model, predictor_name="transformer_qa", cuda_device=args.cuda_device
+    )
+    instances = predictor._dataset_reader.read(args.input_file)
+    logger.info("Running on %d instances", len(instances))
+
+    # We have to make sure we put instances with the same qid all into the same batch.
+    def batch_instances_by_qid(instances: Iterable[Instance]) -> Iterable[List[Instance]]:
+        current_qid = None
+        current_batch = []
+        for instance in instances:
+            instance_qid = instance["metadata"]["id"]
+            if current_qid is None:
+                current_qid = instance_qid
+            if instance_qid == current_qid:
+                current_batch.append(instance)
+            else:
+                yield current_batch
+                current_batch = [instance]
+                current_qid = instance_qid
+        if len(current_batch) > 0:
+            yield current_batch
+
+    def make_batches(
+        instances: Iterable[Instance], batch_size: int = 64
+    ) -> Iterable[List[Instance]]:
+        current_batch = []
+        for qid_instances in batch_instances_by_qid(instances):
+            if len(qid_instances) + len(current_batch) < batch_size:
+                current_batch.extend(qid_instances)
+            else:
+                if len(current_batch) > 0:
+                    yield current_batch
+                current_batch = qid_instances
+        if len(current_batch) > 0:
+            yield current_batch
+
+    # Run model and evaluate results
+    last_logged_scores_time = time.time()
```

Comment:
Not that this will result in a critical failure or anything, but I'd default to `time.monotonic` for this. See https://docs.python.org/3/library/time.html#time.monotonic

Reply:
Neat! 35a033fed05242b17e367ba23f17a218361faac9